usr/share/man/man1/cubic.1
.\" Manual page for cubic.
.\" Copyright (C) 2015, 2020 PJ Singh <psingh.cubic@gmail.com>

.TH cubic 1 "December 1, 2022" "" "Cubic"

.SH NAME
\fBcubic\fR - Custom Ubuntu ISO Creator

.SH SYNOPSIS
\fBcubic\fR [-h] [-v] [-V] [directory] [iso]

.SH DESCRIPTION
Cubic (Custom Ubuntu ISO Creator) is a GUI wizard to create a customized Live
ISO image for Ubuntu and Debian based distributions.
.PP
Cubic permits effortless navigation through the ISO customization steps and
features an integrated virtual command line environment to customize the Linux
file system. You can create new customization projects or modify existing
projects. Important parameters are dynamically populated with intelligent
defaults to simplify the customization process.
.PP
See <https://github.com/PJ-Singh-001/Cubic/wiki> for more information.

.SH OPTIONS

\fBdirectory\fR
.RS
directory for a new or existing project
.RE

\fBiso\fR
.RS
original ISO file for a new project (ignored for existing projects)
.RE

\fB-h, --help\fR
.RS
show the help message and exit
.RE

\fB-v, --verbose\fR
.RS
output formatted log to the console
.RE

\fB-V, --version\fR
.RS
print version information and exit
.RE

.SH NOTES
Cubic is a graphical user interface application and should be started using the
application launcher.
.PP
To run Cubic from the command line, use the \fBcubic\fR command; this executable
is located in the /usr/bin directory and should be available in your path. Do
not run cubic using sudo or as root.

.SH BUGS
Report bugs to <https://github.com/PJ-Singh-001/Cubic/issues>.

.SH AUTHOR
PJ Singh <psingh.cubic@gmail.com>


debian/changelog
cubic (2022.12) bionic focal groovy hirsute impish jammy kinetic lunar; urgency=low

  * Issues fixed in version 2022.12-74:

  - Fixed Bug GH:#130, Error `invalid file name 'boot=casper'` when booting from
    an ISO created with Cubic version 2022.11.73.
  - Enhancement GH:#133, Highlight automatic changes to the boot configuration
    files.
  - Enhancement GH:#129, Include Debian in the Cubic man page description and
    command line help.
  - Enhancement GH:#119, Open the file chooser to the previously used *.iso file
    when selecting an original ISO for new projects. Includes fix to pre-select
    directories.
  - Remove `grml64-full.squashfs` from the predefined list of `squashfs` files
    in the constants module. See Enhancement GH:#93, Add support for Grml Live
    Linux.
  - Revise file_utilities.find_files_with_pattern() to traverse symlinks to
    support Pop_OS! which uses a symlink named `casper` for the *.squashfs
    directory.

  * Issues fixed in version 2022.11-73:

  - Improve text search and replace in boot files.
  - Update file choosers to open the home directory if a file or directory does
    not exist.
  - Automatically identify boot files; remove the list of distro specific boot
    files in the constants module.
  - Enhancement GH:#114, Add support for Debian.
  - Enhancement GH:#93, Add support for Grml Live Linux.
  - Enhancement GH:#116, Do not show the Test button on the Project page and the
    Finish page if the dummy Qemu package (qemu-system-x86 version 0.0) has been
    installed.
  - Enhancement GH:#121, Add auto completion for Cubic command line arguments.
    Add file `/usr/share/bash-completion/completions/cubic`. Require package
    `python3-argcomplete`.
  - Enhancement GH:#117, Format the output of the `cubic --version` command.
  - Enhancement GH:#118, Remove "New" and "★" in the headerbar and menu items.
  - Enhancement GH:#119, Open the file chooser to the previously used *.iso file
    when selecting an original ISO for new projects.
  - Enhancement GH:#120, Introduce an application specific configuration file
    `~/.config/cubic/cubic.conf`. Updated the configuration module: add the
    Configuration base class, the Application derived class, and the Project
    derived class.

  * Issues fixed in version 2022.06-72:

  - Fixed Bug GH:#74, No longer able to query installed software when running
    Cubic in Ubuntu 18.04.5 thru 20.04. Use "dpkg-query --admindir" instead of
    "dpkg-query --root".

  * Issues fixed in version 2022.06-71:

  - Fixed Bug GH:#30, Permission denied error after customization of image is
    complete. Reverted to using dpkg-query, instead of apt.Cache, to retrieve
    package versions and the list of installed packages.

  * Issues fixed in version 2022.01-70:

  - Fixed Bug GH:#19, Unable to locate the squashfs file during customization of
    Ubuntu Server.

  * Issues fixed in version 2022.01-69:

  - Fixed Enhancement GH:#13, Open new Cubic instance with project folder
    pre-selected.
  - Added support for command line arguments, and added the following optional
    command line arguments:
      directory      directory for a new or existing project
      iso            original ISO file for a new project
      -h, --help     show this help message and exit
      -v, --verbose  output formatted log to the console
      -V, --version  print version information and exit
  - Updated the man page to include command line argument instructions.
  - Made logging optional; logging is now off by default.

  * Issues fixed in version 2021.12-68:

  - Fixed Bug GH:#5, Unable to proceed after bash(chroot) step when project
    directory contains nn.nn.n-n format.

  - Fixed Bug GH:#2, Popup Menu and About Dialog warning instance 'XXXXXXXXX'
    has no handler with id 'XXXXX'.

  - Improved processor module to accept a list of arguments for spawned commands
    that require arguments.

  - Added code comments to various modules.

  * Issues fixed in version 2021.12-67:

  - Fixed Bug GH:#1, Migrate the Cubic website from Launchpad to GitHub. (See
    GitHub https://github.com/PJ-Singh-001/Cubic/issues/1).

  * Issues fixed in version 2021.12-66:

  - Fixed Bug LP:#1954558, On the Preseed and Boot tabs, make new files with the
    *.sh extension executable.

  * Issues fixed in version 2021.12-65:

  - Fixed Bug LP:#1954444, Unable to write to file error on typical removable
    packages page.

  * Issues fixed in version 2021.12-64:

  - Improved the progressor module and removed the ProgressTracker thread class.
  - Fixed Bug LP:#1953635, Copying numerous small files is slow.
  - Consolidated file read and write operations to the file_utilities module.
  - Fixed Bug LP:#1951662, On the Copy pages, pause to allow the user to view
    the results before automatically transitioning away from the page.
  - Fixed Bug LP:#1951693, On the Copy pages, clicking the Cancel button may be
    unresponsive.
  - Fixed Bug LP:#1951695, Show error status on the Copy pages.
  - Fixed Bug LP:#1951715, On the Generate page, clicking the Back button during
    update check sums may be unresponsive.
  - Fixed Bug LP:#1952816, Double click or press Enter to select a file in the
    file choosers.

  * Issues fixed in version 2021.11-63:

  - Fixed Bug LP:#1952680, Cannot install Cubic in Linux Mint 19.x.

  * Issues fixed in version 2021.11-62:

  - Fixed Bug LP:#1949237, Can not customize Ubuntu Server 21.10 using Cubic.
    Automatically determine if "ubuntu-server-minimal.squashfs" or
    "filesystem.squashfs" should be extracted. Added "squashfs_file_name"
    parameter to the "cubic.conf" file.
  - Fixed Bug LP:#1950865, The Packages page should only be shown if Ubiquity is
    installed. Added logic to bypass the Packages page if Ubiquity is not
    installed, and added logic to display the minimal install check boxes on the
    Packages page if the version of Ubiquity is >= 18.04. Added python3-apt

usr/share/cubic/cubic.py
#!/usr/bin/python3

########################################################################
#                                                                      #
# cubic.py                                                             #
#                                                                      #
# Copyright (C) 2020 PJ Singh <psingh.cubic@gmail.com>                 #
#                                                                      #
########################################################################

########################################################################
#                                                                      #
# This file is part of Cubic - Custom Ubuntu ISO Creator.              #
#                                                                      #
# Cubic is free software: you can redistribute it and/or modify        #
# it under the terms of the GNU General Public License as published by #
# the Free Software Foundation, either version 3 of the License, or    #
# (at your option) any later version.                                  #
#                                                                      #
# Cubic is distributed in the hope that it will be useful,             #
# but WITHOUT ANY WARRANTY; without even the implied warranty of       #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the         #
# GNU General Public License for more details.                         #
#                                                                      #
# You should have received a copy of the GNU General Public License    #
# along with Cubic. If not, see <http://www.gnu.org/licenses/>.        #
#                                                                      #
########################################################################

########################################################################
# References
########################################################################

# N/A

########################################################################
# Initialize
########################################################################

# import warnings

import os
if os.getuid() == 0:
    print(
        'Error. Cubic (Custom Ubuntu ISO Creator) is a graphical user interface application and may not be run using sudo or as root. See "man cubic" for more information.'
    )
    print()
    exit()

from utilities import logger
logger.log_title('Cubic - Custom Ubuntu ISO Creator')

########################################################################
# Imports
########################################################################

import gi

gi.require_version('Gtk', '3.0')

from gi.repository import Gtk

import glob
import importlib
import os
import traceback

import navigator
from utilities import constructor
from utilities import model

########################################################################
# Global Variables & Constants
########################################################################

# N/A

########################################################################
# Main Application
########################################################################

# 1. Load the UI into the builder from the file.
# 2. Connect the signals to the module that has the handlers.

try:

    logger.log_title('Start Cubic')

    #-------------------------------------------------------------------
    # Initialize
    #-------------------------------------------------------------------

    # Real path is necessary here.
    model.application.directory = os.path.dirname(os.path.realpath(__file__))
    os.chdir(model.application.directory)

    model.application.user_home = os.path.expanduser(os.path.join('~', '*'))

    model.application.cubic_version = constructor.get_package_version('cubic')
    model.application.kernel_version = constructor.get_kernel_version()

    # Load the user interface.
    model.builder = Gtk.Builder.new_from_file('cubic.ui')
    # Connect the signals to handlers in the associated module.
    model.builder.connect_signals(navigator)

    #-------------------------------------------------------------------
    # Pages
    #-------------------------------------------------------------------

    logger.log_label('Setup pages')
    print()

    # Get the stack.
    pages = model.builder.get_object('pages')

    for file_name in sorted(glob.glob('pages/*_page.ui')):

        # Get the module name.
        module_name = file_name[6:-3]
        logger.log_value('Setup', module_name.replace('_', ' '))

        # Load the user interface.
        model.builder.add_from_file(file_name)

        # Load the module.
        module = importlib.import_module('pages.%s' % module_name)

        # Connect the signals to handlers in the associated module.
        model.builder.connect_signals(module)

        # Add page to stack.
        page = model.builder.get_object(module.name)
        pages.add_named(page, module.name)

    # Set the first page for the stack.
    page = model.builder.get_object('start_page')
    pages.set_visible_child(page)

    #-------------------------------------------------------------------
    # Header Bar
    #-------------------------------------------------------------------

    # Add previously loaded widgets to the header bar.

    # TODO: Can adding widgets to the header bar be moved to the individual pages?

    header_bar = model.builder.get_object('header_bar')

    # Project page
    widget = model.builder.get_object('project_page__delete_header_bar_button')
    header_bar.add(widget)

    widget = model.builder.get_object('project_page__header_bar_box')
    header_bar.add(widget)

    # Packages page
    widget = model.builder.get_object('packages_page__header_bar_box')
    header_bar.add(widget)

    # Options page
    options_page__stack_switcher = model.builder.get_object('options_page__stack_switcher')
    options_page__stack = model.builder.get_object('options_page__stack')
    options_page__stack_switcher.set_stack(options_page__stack)

    # Terminal page
    widget = model.builder.get_object('terminal_page__copy_header_bar_button')
    header_bar.add(widget)

    #-------------------------------------------------------------------
    # File Choosers
    #-------------------------------------------------------------------

    logger.log_label('Setup file choosers')

    for file_name in sorted(glob.glob('file_choosers/*_chooser.ui')):

        print()

        # Get the module name.
        module_name = file_name[14:-3]
        logger.log_value('Setup (ignore warnings)', module_name.replace('_', ' '))

        # Load the user interface.
        model.builder.add_from_file(file_name)

        # Load the module.
        module = importlib.import_module('file_choosers.%s' % module_name)

        # Connect the signals to handlers in the associated module.
        model.builder.connect_signals(module)

    #-------------------------------------------------------------------
    # Start the User Interface
    #-------------------------------------------------------------------

    # Show the window.
    window = model.builder.get_object('window')
    window.show()

    # Open the application.
    navigator.handle_navigation('open')

    # Start the Gtk main loop.
    Gtk.main()

except Exception as exception:
    logger.log_value('Exception', exception)
    logger.log_value('The tracek back is', traceback.format_exc())
